src/file.cpp

```diff
@@ -1950,34 +1957,31 @@ typedef struct _FILE_ALLOCATED_RANGE_BUFFER {
 			PTOKEN_PRIVILEGES PreviousState,
 			PDWORD ReturnLength);
 
-		static OpenProcessToken_t pOpenProcessToken = NULL;
-		static LookupPrivilegeValue_t pLookupPrivilegeValue = NULL;
-		static AdjustTokenPrivileges_t pAdjustTokenPrivileges = NULL;
-		static bool failed_advapi = false;
-		HMODULE advapi = NULL;
+		static BOOL result = -1;
+
+		if (result != -1) return !!result;
+
+		result = FALSE;
+		HMODULE advapi = LoadLibraryA("advapi32");
+
+		if (advapi == NULL) return false;
 
 		BOOST_SCOPE_EXIT(&advapi) {
-			if (advapi) FreeLibrary(advapi);
+			FreeLibrary(advapi);
 		} BOOST_SCOPE_EXIT_END
 
-		if (pOpenProcessToken == NULL && !failed_advapi)
+		OpenProcessToken_t const pOpenProcessToken =
+				(OpenProcessToken_t)GetProcAddress(advapi, "OpenProcessToken");
+		LookupPrivilegeValue_t const pLookupPrivilegeValue =
+				(LookupPrivilegeValue_t)GetProcAddress(advapi, "LookupPrivilegeValueA");
+		AdjustTokenPrivileges_t const pAdjustTokenPrivileges =
+				(AdjustTokenPrivileges_t)GetProcAddress(advapi, "AdjustTokenPrivileges");
+
+		if (pOpenProcessToken == NULL
+			|| pLookupPrivilegeValue == NULL
+			|| pAdjustTokenPrivileges == NULL)
 		{
-			advapi = LoadLibraryA("advapi32");
-			if (advapi == NULL)
-			{
-				failed_advapi = true;
-				return false;
-			}
-			pOpenProcessToken = (OpenProcessToken_t)GetProcAddress(advapi, "OpenProcessToken");
-			pLookupPrivilegeValue = (LookupPrivilegeValue_t)GetProcAddress(advapi, "LookupPrivilegeValueA");
-			pAdjustTokenPrivileges = (AdjustTokenPrivileges_t)GetProcAddress(advapi, "AdjustTokenPrivileges");
-			if (pOpenProcessToken == NULL
-				|| pLookupPrivilegeValue == NULL
-				|| pAdjustTokenPrivileges == NULL)
-			{
-				failed_advapi = true;
-				return false;
-			}
```

Comment:
Ah, because `failed_advapi == true` doesn't make the function return with a failure (I think that was the intention).

You may have noticed that this function is quite different in `master`, with a utility function for looking up functions out of DLLs, that's why I'm interested in keeping these changes small and simple. But This issue clearly needs to be fixed.